<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork971-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

-   Acquire data in various ways
-   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="#data_acquisition">Data Acquisition</a>
    <li><a href="#basic_insight">Basic Insight of Dataset</a></li>
</ol>


</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Automobile Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.
<ul>
    <li>Data source: <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork971-2022-01-01" target="_blank">https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


In [ ]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [13]:
# import pandas library
import pandas as pd
import numpy as np

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

Because the data does not include headers, we can add an argument <code>headers = None</code> inside the <code>read_csv()</code> method so that pandas will not automatically set the first row as a header.<br>

You can also assign the dataset to any variable you create.
</p>


This dataset was hosted on IBM Cloud object. Click <a href="https://cocl.us/DA101EN_object_storage">HERE</a> for free storage.


In [14]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"
other_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"
df = pd.read_csv(other_path, header=None)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [ ]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 
df.tail()

0   1      2       3      4     5      6    7      8      9   ...   16  \
200  -1  95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  141   
201  -1  95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   
202  -1  95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  173   
203  -1  95  volvo  diesel  turbo  four  sedan  rwd  front  109.1  ...  145   
204  -1  95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   

       17    18    19    20   21    22  23  24     25  
200  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
201  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
202  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
203   idi  3.01  3.40  23.0  106  4800  26  27  22470  
204  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[5 rows x 26 columns]

<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Add Headers</h3>
<p>
Take a look at our dataset. Pandas automatically set the header with an integer starting from 0.
</p>
<p>
To better describe our data, we can introduce a header. This information is available at:  <a href="https://archive.ics.uci.edu/ml/datasets/Automobile?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork971-2022-01-01" target="_blank">https://archive.ics.uci.edu/ml/datasets/Automobile</a>.
</p>
<p>
Thus, we have to add headers manually.
</p>
<p>
First, we create a list "headers" that include all column names in order.
Then, we use <code>dataframe.columns = headers</code> to replace the headers with the list we created.
</p>


In [ ]:
# create headers list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


 We replace headers and recheck our dataframe:


In [16]:
df.columns = headers
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

We need to replace the "?" symbol with NaN so the dropna() can remove the missing values:


In [18]:
df1=df.replace('?',np.NaN)

We can drop missing values along the column "price" as follows:


In [20]:
df=df1.dropna(subset=["price"], axis=0)
df=df.dropna(subset=['normalized-losses'], axis=0)
df.head(20)

symboling normalized-losses       make fuel-type aspiration num-of-doors  \
3           2               164       audi       gas        std         four   
4           2               164       audi       gas        std         four   
6           1               158       audi       gas        std         four   
8           1               158       audi       gas      turbo         four   
10          2               192        bmw       gas        std          two   
11          0               192        bmw       gas        std         four   
12          0               188        bmw       gas        std          two   
13          0               188        bmw       gas        std         four   
18          2               121  chevrolet       gas        std          two   
19          1                98  chevrolet       gas        std          two   
20          0                81  chevrolet       gas        std         four   
21          1               118      dodge       gas        std          two   
22          1               118      dodge       gas        std          two   
23          1               118      dodge       gas      turbo          two   
24          1               148      dodge       gas        std         four   
25          1               148      dodge       gas        std         four   
26          1               148      dodge       gas        std         four   
27          1               148      dodge       gas      turbo          NaN   
28         -1               110      dodge       gas        std         four   
29          3               145      dodge       gas      turbo          two   

   body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
3       sedan          fwd           front        99.8  ...          109   
4       sedan          4wd           front        99.4  ...          136   
6       sedan          fwd           front       105.8  ...          136   
8       sedan          fwd           front       105.8  ...          131   
10      sedan          rwd           front       101.2  ...          108   
11      sedan          rwd           front       101.2  ...          108   
12      sedan          rwd           front       101.2  ...          164   
13      sedan          rwd           front       101.2  ...          164   
18  hatchback          fwd           front        88.4  ...           61   
19  hatchback          fwd           front        94.5  ...           90   
20      sedan          fwd           front        94.5  ...           90   
21  hatchback          fwd           front        93.7  ...           90   
22  hatchback          fwd           front        93.7  ...           90   
23  hatchback          fwd           front        93.7  ...           98   
24  hatchback          fwd           front        93.7  ...           90   
25      sedan          fwd           front        93.7  ...           90   
26      sedan          fwd           front        93.7  ...           90   
27      sedan          fwd           front        93.7  ...           98   
28      wagon          fwd           front       103.3  ...          122   
29  hatchback          fwd           front        95.9  ...          156   

    fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
3          mpfi  3.19    3.40             10.00        102      5500       24   
4          mpfi  3.19    3.40              8.00        115      5500       18   
6          mpfi  3.19    3.40              8.50        110      5500       19   
8          mpfi  3.13    3.40              8.30        140      5500       17   
10         mpfi  3.50    2.80              8.80        101      5800       23   
11         mpfi  3.50    2.80              8.80        101      5800       23   
12         mpfi  3.31    3.19              9.00        121      4250       21   
13         mpfi  3.31    3.19              9.00        121      4250       21 

Now, we have successfully read the raw dataset and added the correct headers into the dataframe.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [21]:
# Write your code below and press Shift+Enter to execute 
print(df.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>automobile.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


In [23]:
df.to_csv("automobile.csv", index=False)

We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.
</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:
</p>


In [24]:
df.dtypes


symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

A series with the data type of each column is returned.


In [25]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


<p>
As shown above, it is clear to see that the data type of "symboling" and "curb-weight" are <code>int64</code>, "normalized-losses" is <code>object</code>, and "wheel-base" is <code>float64</code>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [26]:
df.describe()

symboling  wheel-base      length       width      height  \
count  164.000000  164.000000  164.000000  164.000000  164.000000   
mean     0.792683   98.164024  172.238415   65.598780   53.774390   
std      1.225874    5.120198   11.417833    1.923028    2.343942   
min     -2.000000   86.600000  141.100000   60.300000   49.400000   
25%      0.000000   94.500000  165.675000   64.000000   52.000000   
50%      1.000000   96.550000  172.000000   65.400000   54.100000   
75%      2.000000  100.400000  177.800000   66.500000   55.500000   
max      3.000000  115.600000  202.600000   71.700000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   164.000000   164.000000         164.000000  164.000000   164.000000  
mean   2458.274390   117.957317          10.126951   26.268293    31.847561  
std     475.087068    30.896294           3.836306    6.193305     6.514349  
min    1488.000000    61.000000           7.000000   15.000000    18.000000  
25%    2090.750000    97.000000           8.700000   22.000000    28.000000  
50%    2367.500000   109.000000           9.000000   26.000000    32.000000  
75%    2785.500000   131.750000           9.400000   31.000000    37.000000  
max    4066.000000   258.000000          23.000000   49.000000    54.000000

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.
<br>

However, what if we would also like to check all the columns including those that are of type object?
<br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [27]:
# describe all the columns in "df"
df.describe(include='all')

symboling normalized-losses    make fuel-type aspiration  \
count   164.000000               164     164       164        164   
unique         NaN                51      18         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      31       149        136   
mean      0.792683               NaN     NaN       NaN        NaN   
std       1.225874               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           163        164          164             164  164.000000  ...   
unique            2          5            3               1         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq             95         80          106             164         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.164024  ...   
std             NaN        NaN          NaN             NaN    5.120198  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   96.550000  ...   
75%             NaN        NaN          NaN             NaN  100.400000  ...   
max             NaN        NaN          NaN             NaN  115.600000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    164.000000          164   160     160        164.000000        164   
unique          NaN            7    33      31               NaN         49   
top             NaN         mpfi  3.62    3.03               NaN         68   
freq            NaN           66    20      14               NaN         18   
mean     117.957317          NaN   NaN     NaN         10.126951        NaN   
std       30.896294          NaN   NaN     NaN          3.836306        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.700000        NaN   
50%      109.000000          NaN   NaN     NaN          9.000000        NaN   
75%      131.750000          NaN   NaN     NaN          9.400000        NaN   
max      258.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        164  164.000000  164.000000   164  
unique        20         NaN         NaN   150  
top         4800         NaN         NaN  7957  
freq          35         NaN         NaN     2  
mean         NaN   26.268293   31.847561   NaN  
std          NaN    6.193305    6.514349   NaN  
min          NaN   15.000000   18.000000   NaN  
25%          NaN   22.000000   28.000000   NaN  
50%          NaN   26.000000   32.000000   NaN  
75%          NaN   31.000000   37.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

In [ ]:
# describe certain the columns in "df" 
df.describe()

symboling  wheel-base      length       width      height  \
count  164.000000  164.000000  164.000000  164.000000  164.000000   
mean     0.792683   98.164024  172.238415   65.598780   53.774390   
std      1.225874    5.120198   11.417833    1.923028    2.343942   
min     -2.000000   86.600000  141.100000   60.300000   49.400000   
25%      0.000000   94.500000  165.675000   64.000000   52.000000   
50%      1.000000   96.550000  172.000000   65.400000   54.100000   
75%      2.000000  100.400000  177.800000   66.500000   55.500000   
max      3.000000  115.600000  202.600000   71.700000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   164.000000   164.000000         164.000000  164.000000   164.000000  
mean   2458.274390   117.957317          10.126951   26.268293    31.847561  
std     475.087068    30.896294           3.836306    6.193305     6.514349  
min    1488.000000    61.000000           7.000000   15.000000    18.000000  
25%    2090.750000    97.000000           8.700000   22.000000    28.000000  
50%    2367.500000   109.000000           9.000000   26.000000    32.000000  
75%    2785.500000   131.750000           9.400000   31.000000    37.000000  
max    4066.000000   258.000000          23.000000   49.000000    54.000000

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>
</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'length' and 'compression-ratio'.

</div>


In [28]:
# Write your code below and press Shift+Enter to execute 
df[['length','compression-ratio']].describe()

length  compression-ratio
count  164.000000         164.000000
mean   172.238415          10.126951
std     11.417833           3.836306
min    141.100000           7.000000
25%    165.675000           8.700000
50%    172.000000           9.000000
75%    177.800000           9.400000
max    202.600000          23.000000


<details><summary>Click here for the solution</summary>

```python
df[['length', 'compression-ratio']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


In [ ]:
dataframe.info()

It provides a concise summary of your DataFrame. 

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [ ]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 164 entries, 3 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          164 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               164 non-null    object 
 3   fuel-type          164 non-null    object 
 4   aspiration         164 non-null    object 
 5   num-of-doors       163 non-null    object 
 6   body-style         164 non-null    object 
 7   drive-wheels       164 non-null    object 
 8   engine-location    164 non-null    object 
 9   wheel-base         164 non-null    float64
 10  length             164 non-null    float64
 11  width              164 non-null    float64
 12  height             164 non-null    float64
 13  curb-weight        164 non-null    int64  
 14  engine-type        164 non-null    object 
 15  num-of-cylinders   164 non-null    object 
 16  engine-size        164 non

<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork971-2022-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork971-2022-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork971-2022-01-01" target="_blank">Fiorella Wenver</a>

<a href=" https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                       |
| ----------------- | ------- | ---------- | ---------------------------------------- |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                   |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'     |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab       |


<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
